File: providers/gpt_image_moderation.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

GPT_IMAGE_MODEL_FAMILIES = (
    "gpt-image-1",
    "gpt-image-1-mini",
    "gpt-image-1.5",
    "gpt-image-2",
)
GPT_IMAGE_MODERATION_VALUES = frozenset({"auto", "low"})


def is_gpt_image_model(model_id: Any) -> bool:
    """Recognize official GPT Image IDs and relay-specific suffixed aliases."""
    normalized = str(model_id or "").strip().lower()
    if not normalized:
        return False
    if ":" in normalized:
        normalized = normalized.split(":", 1)[1]
    basename = normalized.rsplit("/", 1)[-1]
    return any(
        basename == family or basename.startswith(f"{family}-")
        for family in GPT_IMAGE_MODEL_FAMILIES
    )


def apply_gpt_image_moderation_default(
    payload: Mapping[str, Any],
    *,
    model_id: Any = None,
) -> tuple[dict[str, Any], bool]:
    """Default GPT Image generations to low moderation without mutating input."""
    normalized = dict(payload)
    selected_model = model_id if model_id is not None else normalized.get("model")
    if not is_gpt_image_model(selected_model):
        return normalized, False

    moderation = normalized.get("moderation")
    if moderation is None:
        normalized["moderation"] = "low"
        return normalized, True
    if moderation not in GPT_IMAGE_MODERATION_VALUES:
        raise ValueError("GPT Image moderation must be one of: auto, low")
    return normalized, False

File: worker/gpt-image-moderation.mjs
const GPT_IMAGE_MODEL_FAMILIES = Object.freeze([
  "gpt-image-1",
  "gpt-image-1-mini",
  "gpt-image-1.5",
  "gpt-image-2",
]);
const GPT_IMAGE_MODERATION_VALUES = new Set(["auto", "low"]);

export class GPTImageModerationError extends Error {
  constructor() {
    super("GPT Image moderation must be one of: auto, low");
    this.name = "GPTImageModerationError";
  }
}

export function isGptImageModel(modelId) {
  let normalized = typeof modelId === "string" ? modelId.trim().toLowerCase() : "";
  if (!normalized) {
    return false;
  }
  if (normalized.includes(":")) {
    normalized = normalized.split(":", 2)[1];
  }
  const basename = normalized.split("/").at(-1) ?? "";
  return GPT_IMAGE_MODEL_FAMILIES.some(
    (family) => basename === family || basename.startsWith(`${family}-`),
  );
}

export function applyGptImageModerationDefault(payload) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    return { payload, changed: false };
  }
  if (!isGptImageModel(payload.model)) {
    return { payload, changed: false };
  }
  if (payload.moderation != null) {
    if (!GPT_IMAGE_MODERATION_VALUES.has(payload.moderation)) {
      throw new GPTImageModerationError();
    }
    return { payload, changed: false };
  }
  return {
    payload: { ...payload, moderation: "low" },
    changed: true,
  };
}

export async function withDefaultGptImageModeration(request, upstreamPath) {
  const contentType = request.headers.get("content-type")?.toLowerCase() ?? "";
  if (
    request.method !== "POST" ||
    upstreamPath.toLowerCase() !== "/v1/images/generations" ||
    !contentType.startsWith("application/json")
  ) {
    return request;
  }

  let payload;
  try {
    payload = await request.clone().json();
  } catch {
    return request;
  }
  const decision = applyGptImageModerationDefault(payload);
  if (!decision.changed) {
    return request;
  }

  const headers = new Headers(request.headers);
  headers.delete("content-length");
  return new Request(request, {
    headers,
    body: JSON.stringify(decision.payload),
    duplex: "half",
  });
}
